### PR TITLE
Add support for sharding the client reader.

### LINKED ISSUE #7672: Shard Client Reader for registration

### PR DESCRIPTION
**Pull request by clokep**
This is mostly just tests, but does make one change to ensure that a background process only runs on a single process. I'm not 100% sure this is necessary, but I think it is.

Fixes #7672

### CODE REVIEW

File: changelog.d/7830.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Support sharding the client reader workers.
```
Comment (erikjohnston) [inline changelog.d/7830.feature:1]:
```suggestion
Add support for handling registration requests across multiple client reader workers.
```

File: synapse/handlers/deactivate_account.py [hunk 2]
```diff
@@ -40,7 +40,8 @@ def __init__(self, hs):
 
         # Start the user parter loop so it can resume parting users from rooms where
         # it left off (if it has work left to do).
-        hs.get_reactor().callWhenRunning(self._start_user_parting)
+        if hs.config.worker_app is None:
+            hs.get_reactor().callWhenRunning(self._start_user_parting)
```
Comment (clokep) [inline synapse/handlers/deactivate_account.py:44]: I think it is worth pointing out also that we don't seem to use this pattern of `callWhenRunning` anywhere else for background processes. I wonder if that method can just go away and we can directly call the code from it?

Not really related to this PR, just found it odd.

**Comment by clokep** [comment]
@erikjohnston Not sure if you meant to approve this or not. 😄

File: synapse/handlers/deactivate_account.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/handlers/deactivate_account.py:44]: I think we tend to do loopings calls, which also won't run until we're started? I think it makes sense to only start work once we've finished starting up

**Comment by erikjohnston** [comment]
Oh, I was waiting on words in `docs/workers.md`

**Pull request merged**